onnxruntime/test/contrib_ops/ngram_test.cc

```diff
@@ -0,0 +1,395 @@
+const int opset_ver = 1;
+
+void InitTestAttr(OpTester& test, const std::string& mode,
+                  int64_t M, int64_t N, int64_t S, bool all,
+                  const std::vector<int64_t>& ngram_counts,
+                  const std::vector<int64_t>& ngram_indexes,
+                  const std::vector<float>& weights,
+                  const std::vector<int64_t>& pool_int64s,
+                  const std::vector<std::string>& pool_strings) {
+  test.AddAttribute("mode", mode);
+  test.AddAttribute("M", M);
+  test.AddAttribute("N", N);
+  test.AddAttribute("S", S);
+  test.AddAttribute("all", int64_t{all});
+  test.AddAttribute("ngram_counts", ngram_counts);
+  test.AddAttribute("ngram_indexes", ngram_indexes);
+  // optional
+  if (!weights.empty()) {
+    test.AddAttribute("weights", weights);
+  }
+  if (!pool_int64s.empty()) {
+    test.AddAttribute("pool_int64s", pool_int64s);
+  } else {
+    test.AddAttribute("pool_strings", pool_strings);
+  }
+}
+}  // namespace ngram_test
+
+using namespace ngram_test;
+
+TEST(ContribOpTest, Ngram_Int32_TF_AllFalse_Skip0) {
+  OpTester test("Ngram", opset_ver, domain);
+  // 1 - 2, s=0, , all=false, weights empty, int32
+  InitTestAttr(test, "TF", 1, 2, 0, false,
+               {0, 4},
+               {0, 1, 2, 3, 4, 5, 6},  //7 output indexes
+               {},
+               {2, 3, 5, 4,         //1-grams
+                5, 6, 7, 8, 6, 7},  //bi-grams
+               {});
+
+  std::vector<int64_t> dims{2, 6};
+  std::vector<int32_t> input = {1, 1, 3, 3, 3, 7, 8, 6, 7, 5, 6, 8};
+  test.AddInput<int32_t>("T", dims, input);
+
+  std::vector<int64_t> out_dims{7};
+  std::vector<float> output = {0, 0, 0, 0, 1, 1, 1};
+  test.AddOutput<float>("Y", out_dims, output);
+
+  test.Run(OpTester::ExpectResult::kExpectSuccess);
+}
+
+TEST(ContribOpTest, Ngram_String_TF_AllFalse_Skip0) {
+  OpTester test("Ngram", opset_ver, domain);
+  // 1 - 2, s=0, , all=false, weights empty, string
+  InitTestAttr(test, "TF", 1, 2, 0, false,
+               {0, 4},
+               {0, 1, 2, 3, 4, 5, 6},  //7 output indexes
+               {},
+               {},
+               {"two", "three", "five", "four",                     //1-grams
+                "five", "six", "seven", "eight", "six", "seven"});  //bi-grams
+
+  std::vector<int64_t> dims{2, 6};
+  std::vector<std::string> input{"one", "one", "three", "three", "three", "seven", "eight",
+                                 "six", "seven", "five", "six", "eight"};
+  test.AddInput<std::string>("T", dims, input);
+
+  std::vector<int64_t> out_dims{7};
+  std::vector<float> output = {0, 0, 0, 0, 1, 1, 1};
+  test.AddOutput<float>("Y", out_dims, output);
+
+  test.Run(OpTester::ExpectResult::kExpectSuccess);
+}
+
+TEST(ContribOpTest, Ngram_Int32_TF_AllFalse_Skip0_LevelEmpty) {
+  OpTester test("Ngram", opset_ver, domain);
+  // 1 - 2, s=0, , all=false, weights empty, int32
+  InitTestAttr(test, "TF", 1, 2, 0, false,
+               {0, 0},
```

Comment:
>0 [](start = 16, length = 1)

Would it be clearer to use -1 to indicate no entries of a specific length?

Reply:
This is a starting index not the length. The length is computed by computing a difference of two adjacent indices per spec. Having -1 would introduce a code branch but I do not have a strong opinion on this and will let @wschin@outlook.com To chime in.

---
In reply to: [242803205](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r242803205) [](ancestors = 242803205)

Reply:
OK. Would be good to add a comment somewhere pointing out the similarity to CSR so that someone who doesn't know that format (like me) can go look it up and have it make more sense.

---
In reply to: [243122883](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243122883) [](ancestors = 243122883,243107449,242803205)